### ISSUE TITLE
Multi Layer Perceptron Classifier predicting the test data set to a single label

### ISSUE THREAD

**Issue by raunaq1990** [haifengl/smile#550]
First of all, I am fairly new to ML programming. I have some experience in Scala, and based on recommendations in internet, i decided to try out SMILE.
I seem to be running into a problem where my ANN model is always predicting a single label value for my entire test dataset.
I am using a data set with 96 features, ~20k samples for training, ~7.5k samples for testing, and 8 distinct output classes.
I have created my MLP model as follows
`val mlp = new MLP(96, Layer.sigmoid(128), Layer.sigmoid(64), Layer.sigmoid(32), Layer.mle(8, OutputFunction.SOFTMAX))`
Then i divided my training data set into mini batches of size 32.
`mlp.update(batch, batchLabel)`
Then i am doing the prediction as follows
`    val pred = new Array[Int](testLabels.size)
    for (i <- 0 until testLabels.size) {
      pred(i) = mlp.predict(testData(i))
    }
`
I have compared the actual test labels and the predicted test labels, and the count per distinct labels are:
`Actual count per label: 879, 1205, 145, 517, 1299, 1790, 246, 1380
Predicted count per label: 0, 0, 0, 0, 0, 7461, 0, 0`
I have tried out multiple batch sizes, as well as multiple hidden layer sizes as well, but i always seem to get the same result
My data is also not skewed and contains fair distribution of test data across all 8 labels

**Comment by haifengl** [comment]
It looks like that some weights become NaN during the training. A couple of suggestions:

- It is not a good idea to use sigmoid with many layers. Try ReLU. If your prefer sigmoid, try only one hidden layer.
- The output objective function `mle` is not preferred for classification. Try cross entropy.
- For your problem, try different hyper-parameters like learning rate, momentum, weight decay. Mini batch size and # of hidden units should not have critical connections to your problem.

Good luck!

**Comment by raunaq1990** [comment]
Thanks for the quick response!
I will try out your suggestion and share my results.

**Comment by raunaq1990** [comment]
I tried modifying my model based on your suggestions, and tried with 2 ReLU hidden layers and a Sigmoid activation function with MSE.
`val mlp = new MLP(96, Layer.rectifier(100), Layer.rectifier(50), Layer.mse(8, OutputFunction.SIGMOID))`

I also tried varying the hyper-parameters:
- eta [0.01, 0.001, 0.0001]
- alpha [0.01]
- lambda [0.01, 0.001]

In addition, i also tried normalizing my data.
But the result is still the same as earlier.

Can you provide any further suggestions or modifications?
Also, i wanted to try out Grid Search for hyper-parameter tuning. Any examples on how to do it?

**Comment by haifengl** [comment]
Data representation may be an issue. How do you normalize the data?

For grid search, see #219

**Comment by albertmolinermrf** [comment]
A way to confirm suspicions on normalization is to evaluate the test set before any training is performed. The distribution should be apparently random. If a single label is still getting all predictions, a clear cause could be unnormalized data.

Is the training set shuffled? If so and the predicted label is not always the same, chances are that the learning rate is still too large.

To me, `Layer.mle(8, OutputFunction.SOFTMAX)` was alright for classification.
